File: Frontend/angular-app/src/app/modules/accounting/repositories/moadian/moadian-invoice-header/commands/update-moadian-invoice-header-command.ts
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../../core/models/service-result";
import {IRequest, IRequestHandler} from "../../../../../../core/services/mediator/interfaces";
import {MoadianInvoiceHeader} from "../../../../entities/moadian-invoice-header";
import {HttpService} from "../../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../../core/services/mediator/decorator";
import {ValidationRule} from "../../../../../../core/validation/validation-rule";
import {HttpRequest} from "../../../../../../core/services/http/http-request";
import {
  CreateMoadianInvoiceDetailCommand
} from "../../moadian-invoice-detail/commands/create-moadian-invoice-detail-command";
import {
  UpdateMoadianInvoiceDetailCommand
} from "../../moadian-invoice-detail/commands/update-moadian-invoice-detail-command";
import {NotificationService} from "../../../../../../shared/services/notification/notification.service";

export class UpdateMoadianInvoiceHeaderCommand extends IRequest<UpdateMoadianInvoiceHeaderCommand, MoadianInvoiceHeader> {

  public id: number | undefined = undefined;

  public isSandbox: boolean | undefined = undefined;
  public personFullName: string | undefined = undefined;
  public accountReferenceCode: string | undefined = undefined;
  public customerCode: string | undefined = undefined;

  public errors: string | undefined = undefined;
  public status: string| undefined = undefined;
  public referenceId: string | undefined = undefined;
  public submissionDate: Date | undefined = undefined;

  public taxId: string | undefined = undefined;
  public indatim: number | undefined = undefined;
  public indati2m: number | undefined = undefined;
  public inty: number | undefined = undefined;
  public inno: string | undefined = undefined;
  public invoiceNumber: string | undefined = undefined;
  public irtaxid: string | undefined = undefined;
  public inp: number | undefined = undefined;
  public ins: number | undefined = undefined;
  public tins: string | undefined = undefined;
  public tob: number | undefined = undefined;
  public bid: string | undefined = undefined;
  public tinb: string | undefined = undefined;
  public sbc: string | undefined = undefined;
  public bpc: string | undefined = undefined;
  public bbc: string | undefined = undefined;
  public ft: number | undefined = undefined;
  public bpn: string | undefined = undefined;
  public scln: string | undefined = undefined;
  public scc: string | undefined = undefined;
  public crn: string | undefined = undefined;
  public billid: string | undefined = undefined;
  public tprdis: number | undefined = undefined;
  public tdis: number | undefined = undefined;
  public tadis: number | undefined = undefined;
  public tvam: number | undefined = undefined;
  public todam: number | undefined = undefined;
  public tbill: number | undefined = undefined;
  public setm: number | undefined = undefined;
  public cap: number | undefined = undefined;
  public insp: number | undefined = undefined;
  public tvop: number | undefined = undefined;
  public tax17: number | undefined = undefined;
  public cdcn: string | undefined = undefined;
  public cdcd: number | undefined = undefined;
  public tonw: number | undefined = undefined;
  public torv: number | undefined = undefined;
  public tocv: number | undefined = undefined;

  public moadianInvoiceDetails: UpdateMoadianInvoiceDetailCommand[] = []

  constructor() {
    super();
  }

  mapFrom(entity: MoadianInvoiceHeader): UpdateMoadianInvoiceHeaderCommand {
    this.mapBasics(entity,this);
    this.moadianInvoiceDetails = entity.moadianInvoiceDetails.map(x => new UpdateMoadianInvoiceDetailCommand().mapFrom(x))
    return this;
  }

  mapTo(): MoadianInvoiceHeader {
    throw new ApplicationError(UpdateMoadianInvoiceHeaderCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/accounting/moadian/updateInvoice";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(UpdateMoadianInvoiceHeaderCommandHandler.name)
export class UpdateMoadianInvoiceHeaderCommandHandler implements IRequestHandler<UpdateMoadianInvoiceHeaderCommand, MoadianInvoiceHeader> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService
  ) {
  }

  async Handle(request: UpdateMoadianInvoiceHeaderCommand): Promise<MoadianInvoiceHeader> {
    let httpRequest: HttpRequest<UpdateMoadianInvoiceHeaderCommand> = new HttpRequest<UpdateMoadianInvoiceHeaderCommand>(request.url, request);
    // httpRequest.Query  += ;
    // httpRequest.Headers = httpRequest.Headers.append()

    return await this._httpService.Put<UpdateMoadianInvoiceHeaderCommand, ServiceResult<MoadianInvoiceHeader>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })
  }
}

File: Backend/Microservices/Accounting/Eefa.Accounting.Application.Refactor/UseCases/Moadians/MoadianInvoiceHeader/Commands/Update/UpdateMoadianInvoiceHeaderCommand.cs
using AutoMapper;
using MediatR;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Query;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

public class UpdateMoadianInvoiceHeaderCommand : IRequest<ServiceResult<MoadianInvoiceHeaderDetailedModel>>, IMapFrom<UpdateMoadianInvoiceHeaderCommand>
{
    public int Id { get; set; }
    public string TaxId { get; set; } = default!;
    public long Indatim { get; set; } = default!;
    public long Indati2m { get; set; } = default!;
    public int Inty { get; set; } = default!;
    public string Inno { get; set; } = default!;
    public string? Irtaxid { get; set; }
    public int Inp { get; set; } = default!;
    public int Ins { get; set; } = default!;
    public string Tins { get; set; } = default!;
    public int? Tob { get; set; }
    public string? Bid { get; set; }
    public string? Tinb { get; set; }
    public string? Sbc { get; set; }
    public string? Bpc { get; set; }
    public string? Bbc { get; set; }
    public int? Ft { get; set; }
    public string? Bpn { get; set; }
    public string? Scln { get; set; }
    public string? Scc { get; set; }
    public string? Crn { get; set; }
    public string? Billid { get; set; }
    public decimal Tprdis { get; set; } = default!;
    public decimal Tdis { get; set; } = default!;
    public decimal Tadis { get; set; } = default!;
    public decimal Tvam { get; set; } = default!;
    public decimal Todam { get; set; } = default!;
    public decimal Tbill { get; set; } = default!;
    public int Setm { get; set; } = default!;
    public decimal Cap { get; set; } = default!;
    public decimal Insp { get; set; } = default!;
    public decimal Tvop { get; set; } = default!;
    public decimal Tax17 { get; set; } = default!;
    public string? Cdcn { get; set; }
    public int? Cdcd { get; set; }
    public decimal Tonw { get; set; } = default!;
    public decimal Torv { get; set; } = default!;
    public decimal Tocv { get; set; } = default!;
    public string? Status { get; set; } = default!;

    public List<UpdateMoadianInvoiceDetailCommand> MoadianInvoiceDetails { get; set; } = default!;
    public void Mapping(Profile profile)
    {
        profile.CreateMap<UpdateMoadianInvoiceHeaderCommand, MoadianInvoiceHeader>()
            .IgnoreAllNonExisting();
    }
}


public class UpdateMoadianInvoiceHeaderCommandHandler : IRequestHandler<UpdateMoadianInvoiceHeaderCommand, ServiceResult<MoadianInvoiceHeaderDetailedModel>>
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly IMapper _mapper;
    public UpdateMoadianInvoiceHeaderCommandHandler(IUnitOfWork unitOfWork, IMapper mapper)
    {
        _unitOfWork = unitOfWork;
        _mapper = mapper;
    }
    public async Task<ServiceResult<MoadianInvoiceHeaderDetailedModel>> Handle(UpdateMoadianInvoiceHeaderCommand request, CancellationToken cancellationToken)
    {
        MoadianInvoiceHeader entity = await _unitOfWork.MoadianInvoiceHeaders
                    .GetByIdAsync(request.Id, x => x.Include(y => y.MoadianInvoiceDetails));

        if (entity.IsSandbox || (!entity.IsSandbox && entity.Status != "SUCCESS"))
        {
            _mapper.Map<UpdateMoadianInvoiceHeaderCommand, MoadianInvoiceHeader>(request, entity);
            _unitOfWork.MoadianInvoiceHeaders.Update(entity);

            await _unitOfWork.SaveChangesAsync(cancellationToken);
        }
        else
        {
            throw new Exception("صورتحساب های نهایی قابل ویرایش نمیباشند");
        }

        return ServiceResult.Success(_mapper.Map<MoadianInvoiceHeaderDetailedModel>(entity));
    }
}